draco/analysis/powerspec.py

```diff
@@ -0,0 +1,1504 @@
+    """
+    # Find the Fourier norm for the FFT
+    # The norm is mentioned here - https://numpy.org/doc/2.0/reference/routines.fft.html
+    FT_norm = 1 / float(np.prod(np.array(data.shape)))
+
+    if window:
+        x_ra = (ra - ra[0]) / (ra[-1] - ra[0])
+        x_dec = (dec - dec[0]) / (dec[-1] - dec[0])
+        w_ra = tools.window_generalised(x_ra, window=window)
+        w_dec = tools.window_generalised(x_dec, window=window)
+
+        # estimate the equivalent noise bandwidth for the tapering window
+        NEB_ra = noise_equivalent_bandwidth(ra.size, window)
+        NEB_dec = noise_equivalent_bandwidth(dec.size, window)
+        taper_window = np.outer(w_ra[:, np.newaxis], w_dec[np.newaxis, :])
+        data *= taper_window
+        uv_map = np.fft.fftshift(np.fft.fft2(data))
+
+    else:
+        uv_map = np.fft.fftshift(np.fft.fft2(data))
+        NEB_ra = NEB_dec = 1.0
+
+    return uv_map * FT_norm, NEB_ra, NEB_dec
+
+
+def vol_normalization(ra, dec, freq, redshift, cosmo=None):
+    """Estimate the volume normalization factor for the power spectrum.
+
+    Parameters
+    ----------
+    ra : np.array[ra]
+      The RA array in deg unit.
+    dec : np.narray[dec]
+      The DEC array in deg unit.
+    freq : np.array[freq]
+      The freq array in MHz.
+    redshift : float
+      redshift at the centre of the band.
+    cosmo: Cosmology object
+        Default is cora.util.Cosmology() default.
+
+    Returns
+    -------
+    norm : float
+      The  Ppower spectrum normalization factor in Mpc^3 unit
+    """
+    if cosmo is None:
+        cosmo = get_cosmo()
+
+    nra = ra.size
+    ndec = dec.size
+    nfreq = freq.size
+
+    # Mean resolution in RA and DEC and convert that to radian
+    res_ra_radian = np.deg2rad(np.mean(np.diff(ra)))
+    res_dec_radian = np.deg2rad(np.mean(np.diff(dec)))
+
+    # Comoving distance
+    DMz = dRperp_dtheta(redshift, cosmo=cosmo)  # in [h^-1 Mpc]
+
+    # Convert the RA and DEC resolution to Mpc unit
+    dx_Mpc = DMz * res_ra_radian
+    dy_Mpc = DMz * res_dec_radian
+    Lx = nra * dx_Mpc  # survey length along RA [h^-1 Mpc]
+    Ly = ndec * dy_Mpc  # survey length along DEC [h^-1 Mpc]
+
+    ## Along line-of-sight direction
+    chan_width = np.abs(np.diff(freq)).mean() * 1e6  # channel width in Hz
+    dz_Mpc = dRpara_df(redshift, cosmo=cosmo) * chan_width  # [h^-1 Mpc]
+    Lz = dz_Mpc * nfreq  # survey length along line-of-sight [Mpc]
+
+    return Lx * Ly * Lz
+
+
+def nanaverage(d, w, axis=None):
+    """Estimate the nanaverage data using the weight.
+
+    Parameters
+    ----------
+    d : np.ndarray
+     The data to average
+    w : np.ndarray
+     The weight to use during averaging.
+    axis: int, optional
+     Axis along which the average will be taken.
+
+    Returns
+    -------
+    d_avg : np.ndarray
+     The weighted average.
+    """
+    return np.nansum(d * w, axis=axis) / np.nansum(w, axis=axis)
```

Comment:
I think this will be incorrect unless the NaNs exist in the same indices in both `d` and `w`, or the weights are zeroed wherever NaNs exist in `d`. That may be the case wherever it's used, but I'd think about switching to using the `where` keyword in `sum` so you get something like
```python
np.sum(d * w, axis=axis, where=~np.isnan(d))
```